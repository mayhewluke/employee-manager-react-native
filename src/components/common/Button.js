import React from "react";
import { TouchableOpacity, Text } from "react-native";
import PropTypes from "prop-types";

const styles = {
  buttonStyle: {
    flex: 1,
    alignSelf: "stretch",
    backgroundColor: "#fff",
    borderRadius: 5,
    borderWidth: 1,
    borderColor: "#007aff",
    marginLeft: 5,
    marginRight: 5
  },
  textStyle: {
    alignSelf: "center",
    color: "#007aff",
    fontSize: 16,
    fontWeight: "600",
    paddingTop: 10,
    paddingBottom: 10
  }
};

const Button = ({ onPress, children }) => {
  const { buttonStyle, textStyle } = styles;
  return (
    <TouchableOpacity style={buttonStyle} onPress={onPress}>
      <Text style={textStyle}>{children}</Text>
    </TouchableOpacity>
  );
};

Button.propTypes = {
  onPress: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired
};

export { Button };
