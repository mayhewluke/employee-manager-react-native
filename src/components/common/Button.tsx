import React from "react";
import { StyleSheet, Text, TouchableOpacity } from "react-native";

const styles = StyleSheet.create({
  buttonStyle: {
    backgroundColor: "#fff",
    borderColor: "#007aff",
    borderRadius: 5,
    borderWidth: 1,
    marginLeft: 5,
    marginRight: 5,
  },
  textStyle: {
    alignSelf: "center",
    color: "#007aff",
    fontSize: 16,
    fontWeight: "600",
    paddingBottom: 10,
    paddingTop: 10,
  },
});

const Button: React.SFC<{ onPress: () => void }> = ({ onPress, children }) => {
  const { buttonStyle, textStyle } = styles;
  return (
    <TouchableOpacity style={buttonStyle} onPress={onPress}>
      <Text style={textStyle}>{children}</Text>
    </TouchableOpacity>
  );
};

export default Button;
