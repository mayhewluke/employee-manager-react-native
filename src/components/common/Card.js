import React from "react";
import { View } from "react-native";
import PropTypes from "prop-types";

const styles = {
  containerStyle: {
    borderWidth: 1,
    borderRadius: 2,
    borderColor: "#ddd",
    borderBottomWidth: 0,
    elevation: 1,
    marginLeft: 5,
    marginRight: 5,
    marginTop: 10
  }
};

const Card = ({ children }) => {
  const { containerStyle } = styles;
  return <View style={containerStyle}>{children}</View>;
};

Card.propTypes = {
  children: PropTypes.node.isRequired
};

export default Card;
