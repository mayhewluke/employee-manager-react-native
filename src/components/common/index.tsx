export { default as AsyncButton } from "./AsyncButton";
export { default as Button } from "./Button";
export { default as Card } from "./Card";
export { default as CardSection } from "./CardSection";
export { default as Confirm } from "./Confirm";
export { default as Input } from "./Input";
export { default as Spinner, Size as SpinnerSize } from "./Spinner";
