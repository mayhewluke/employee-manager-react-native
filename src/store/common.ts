export type Async<T> =
  | { state: "INIT" }
  | { state: "PROGRESS" }
  | { state: "COMPLETE"; value: T }
  | { state: "ERROR"; error: string };
