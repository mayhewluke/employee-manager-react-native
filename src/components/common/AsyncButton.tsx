import * as React from "react";
import { StyleSheet, Text, View } from "react-native";

import { Async } from "../../store";
import Button from "./Button";
import Spinner, { Size as SpinnerSize } from "./Spinner";

interface IProps {
  label: string;
  asyncAction: Async<any>;
  onPress: () => void;
  size?: SpinnerSize;
}

const styles = StyleSheet.create({
  errorTextStyle: {
    alignSelf: "center",
    color: "red",
    fontSize: 20,
  },
});

const AsyncButton: React.SFC<IProps> = ({
  label,
  asyncAction,
  onPress,
  size = SpinnerSize.Large,
}) => (
  <View style={{ flex: 1, flexDirection: "row" }}>
    {asyncAction.state === "ERROR" && (
      <Text style={styles.errorTextStyle}>{asyncAction.error}</Text>
    )}
    {asyncAction.state === "PROGRESS" ? (
      <Spinner size={size} />
    ) : (
      <Button onPress={onPress}>{label}</Button>
    )}
  </View>
);

export default AsyncButton;
