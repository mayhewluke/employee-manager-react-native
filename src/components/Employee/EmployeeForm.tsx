import * as React from "react";
import { Picker, StyleSheet, Text, View } from "react-native";
import { connect } from "react-redux";

import { IRootState } from "../../store";
import {
  FieldUpdatePayload,
  ShiftDay,
  updateField,
} from "../../store/Employee";
import { CardSection, Input } from "../common";

const styles = StyleSheet.create({
  errorTextStyle: {
    alignSelf: "center",
    color: "red",
    fontSize: 20,
  },
  pickerContainerStyle: {
    flexDirection: "column",
    paddingLeft: 25,
  },
  pickerLabelStyle: {
    fontSize: 18,
  },
});

export interface IFormProps {
  employeeName: string;
  phone: string;
  shift: ShiftDay;
  dispatchFieldUpdate: (payload: FieldUpdatePayload) => void;
  error: string;
}

const Form = ({
  employeeName,
  phone,
  shift,
  dispatchFieldUpdate,
  error,
}: IFormProps) => {
  const dispatchUpdateName = (value: string) =>
    dispatchFieldUpdate({ value, field: "employeeName" });
  const dispatchUpdatePhone = (value: string) =>
    dispatchFieldUpdate({ value, field: "phone" });
  const dispatchUpdateShift = (value: ShiftDay) =>
    dispatchFieldUpdate({ value, field: "shift" });
  return (
    <View>
      <CardSection>
        <Input
          label="Name"
          placeholder="Taylor"
          value={employeeName}
          onChangeText={dispatchUpdateName}
        />
      </CardSection>
      <CardSection>
        <Input
          label="Phone"
          placeholder="555-555-5555"
          value={phone}
          onChangeText={dispatchUpdatePhone}
        />
      </CardSection>

      <CardSection style={styles.pickerContainerStyle}>
        <Text style={styles.pickerLabelStyle}>Shift day</Text>
        <Picker selectedValue={shift} onValueChange={dispatchUpdateShift}>
          {Object.keys(ShiftDay).map(key => (
            <Picker.Item
              key={key}
              value={key}
              label={ShiftDay[key as keyof typeof ShiftDay]}
            />
          ))}
        </Picker>
      </CardSection>

      {error !== "" && (
        <CardSection>
          <Text style={styles.errorTextStyle}>{error}</Text>
        </CardSection>
      )}
    </View>
  );
};

const mapStateToProps = (state: IRootState) => {
  const { employeeName, phone, shift } = state.employee;
  return { employeeName, phone, shift };
};

export default connect(
  mapStateToProps,
  { dispatchFieldUpdate: updateField },
)(Form);
