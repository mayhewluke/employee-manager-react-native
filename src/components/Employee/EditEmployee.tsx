import React, { Component } from "react";
import { ScrollView, View } from "react-native";
import reactNativeCommunications from "react-native-communications";
import { NavigationScreenProp, NavigationScreenProps } from "react-navigation";
import { connect } from "react-redux";

import { AsyncValue, IRootState } from "../../store";
import {
  closeFireModal,
  fireEmployee,
  IEmployee,
  resetForm,
  ShiftDay,
  showFireModal,
  updateEmployee,
  updateField,
} from "../../store/Employee";
import { AsyncButton, Button, Card, CardSection, Confirm } from "../common";
import EmployeeForm, { IFormProps } from "./EmployeeForm";

interface IProps extends IFormProps, NavigationScreenProps {
  employeeName: string;
  phone: string;
  shift: ShiftDay;
  uid: string | null;
  fireAction: AsyncValue<null>;
  fireModalShown: boolean;
  updateAction: AsyncValue<null>;
  dispatchShowFireModal: () => void;
  dispatchUpdateEmployee: (employee: IEmployee<string>) => void;
  dispatchResetForm: () => void;
  dispatchCloseModal: () => void;
  dispatchFireEmployee: (
    uid: string | null,
    navigation: NavigationScreenProp<any>,
  ) => void;
}

const textSchedule = (phone: string, shift: ShiftDay) => () =>
  reactNativeCommunications.text(phone, `Your upcoming shift is on ${shift}`);

class EditEmployee extends Component<IProps> {
  public static navigationOptions = ({
    navigation,
  }: {
    navigation: NavigationScreenProp<any>;
  }) => ({
    title: `Edit ${navigation.getParam("employeeName")}`,
  });

  public componentWillUnmount() {
    const { dispatchResetForm } = this.props;
    dispatchResetForm();
  }

  public render() {
    const {
      employeeName,
      phone,
      shift,
      uid,
      fireAction,
      fireModalShown,
      updateAction,
      dispatchShowFireModal,
      dispatchUpdateEmployee,
      dispatchCloseModal,
      dispatchFireEmployee,
      navigation,
    } = this.props;
    // Short-circuit and redirect if the employee to edit has not actually been
    // saved yet (can't edit something non-persisted)
    if (uid === null) {
      navigation.navigate("EmployeeList");
      return null;
    }
    const onFireEmployee = () => dispatchFireEmployee(uid, navigation);
    const onUpdateEmployee = () =>
      dispatchUpdateEmployee({
        employeeName,
        phone,
        shift,
        uid,
      });
    return (
      <ScrollView>
        <Card>
          <EmployeeForm {...this.props} />

          <CardSection>
            <AsyncButton
              label="Save"
              asyncAction={updateAction}
              onPress={onUpdateEmployee}
            />
          </CardSection>

          <CardSection>
            <View style={{ flex: 1 }}>
              <Button onPress={textSchedule(phone, shift)}>
                Text schedule
              </Button>
            </View>
          </CardSection>

          <CardSection>
            <AsyncButton
              label="Fire"
              asyncAction={fireAction}
              onPress={dispatchShowFireModal}
            />
          </CardSection>
          <Confirm
            text={`Are you sure you want to fire ${employeeName}?`}
            isVisible={fireModalShown}
            onYes={onFireEmployee}
            onNo={dispatchCloseModal}
          />
        </Card>
      </ScrollView>
    );
  }
}

const mapStateToProps = (state: IRootState) => {
  const {
    employeeName,
    phone,
    shift,
    uid,
    fireModalShown,
    fireAction,
    updateAction,
  } = state.employee;
  return {
    employeeName,
    fireAction,
    fireModalShown,
    phone,
    shift,
    uid,
    updateAction,
  };
};

export default connect(
  mapStateToProps,
  {
    dispatchCloseModal: closeFireModal,
    dispatchFieldUpdate: updateField,
    dispatchFireEmployee: fireEmployee,
    dispatchResetForm: resetForm,
    dispatchShowFireModal: showFireModal,
    dispatchUpdateEmployee: updateEmployee,
  },
)(EditEmployee);
