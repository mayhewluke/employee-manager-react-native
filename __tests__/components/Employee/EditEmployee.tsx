import "jest-enzyme";
import React from "react";

import { shallow, ShallowWrapper } from "enzyme";
import { ScrollView } from "react-native";
import reactNativeCommunications from "react-native-communications";
import { NavigationScreenProps } from "react-navigation";

import { stubNavigation } from "../../helpers/react-navigation";
import { StubbedStore } from "../../helpers/redux";

import { AsyncButton } from "../../../src/components/common";
import EditEmployee from "../../../src/components/Employee/EditEmployee";
import EmployeeForm from "../../../src/components/Employee/EmployeeForm";
import {
  closeFireModal,
  fireEmployee,
  IEmployeeState,
  resetForm,
  ShiftDay,
  showFireModal,
  updateEmployee,
  updateField,
} from "../../../src/store/Employee";

jest.mock("../../../src/store/Employee");
jest.mock(
  "../../../src/components/Employee/EmployeeForm",
  () => "EmployeeForm",
);

// For some reason TypeScript doesn't know about a `connect`ed class' static
// methods, and no solution is readily available after searching, so just
// override the types here until a better solution can be found.
const { navigationOptions } = EditEmployee as any;

describe("EmployeeList", () => {
  let props: NavigationScreenProps;
  let store: StubbedStore;
  let state: { employee: IEmployeeState };
  beforeEach(() => {
    state = {
      employee: {
        createAction: { state: "INIT" },
        employeeName: "Taylor",
        fireAction: { state: "INIT" },
        fireModalShown: false,
        phone: "555-5555",
        shift: ShiftDay.Monday,
        uid: "uid1",
        updateAction: { state: "INIT" },
      },
    };
    store = new StubbedStore(state);
    props = {
      navigation: stubNavigation(),
    };
  });
  // This has side effects! Don't put this in beforeEach in case a test needs to
  // do something *before* running those side effects!
  const mkWrapper = () =>
    shallow(<EditEmployee {...props} />, {
      context: { store },
    }).dive();

  it("is scrollable", () => {
    expect(mkWrapper().type()).toBe(ScrollView);
  });

  it("passes dispatchFieldUpdate to the EmployeeForm", () => {
    // Need to use the actual component type rather than a string or we lose
    // type safety
    const employeeForm = mkWrapper().find(EmployeeForm);
    // Can't just check the functions are equal since the dispatch* version is
    // wrapped inside another function in order to hook up the dispatch
    expect(updateField).toHaveBeenCalledTimes(0);

    employeeForm.props().dispatchFieldUpdate();
    expect(updateField).toHaveBeenCalledTimes(1);
  });

  describe("save button", () => {
    let button: ShallowWrapper;
    beforeEach(() => {
      button = mkWrapper().find({ label: "Save" });
    });

    it("uses an AsyncButton connected to updateAction", () => {
      expect(button.type()).toBe(AsyncButton);
      expect(button.prop("asyncAction")).toBe(state.employee.updateAction);
    });

    it("updates the Employee when pressed", () => {
      const { employeeName, phone, shift, uid } = state.employee;

      expect(updateEmployee).not.toHaveBeenCalled();
      button.simulate("press");

      expect(updateEmployee).toHaveBeenCalledTimes(1);
      expect(updateEmployee).toHaveBeenCalledWith(
        { employeeName, phone, shift, uid },
        props.navigation,
      );
    });
  });

  describe("text schedule button", () => {
    it("creates a text in the default texting app", () => {
      reactNativeCommunications.text = jest.fn();
      const button = mkWrapper().find("Button");
      const { phone, shift } = state.employee;
      expect(reactNativeCommunications.text).not.toHaveBeenCalled();

      button.simulate("press");

      expect(reactNativeCommunications.text).toHaveBeenCalledTimes(1);
      expect(reactNativeCommunications.text).toHaveBeenCalledWith(
        phone,
        expect.stringContaining(shift),
      );
    });
  });

  describe("fire button", () => {
    let button: ShallowWrapper;
    beforeEach(() => {
      button = mkWrapper().find({ label: "Fire" });
    });

    it("uses an AsyncButton connected to fireAction", () => {
      expect(button.type()).toBe(AsyncButton);
      expect(button.prop("asyncAction")).toBe(state.employee.fireAction);
    });

    it("shows the confirmation modal without firing the Employee", () => {
      expect(showFireModal).not.toHaveBeenCalled();
      button.simulate("press");

      expect(fireEmployee).not.toHaveBeenCalled();
      expect(showFireModal).toHaveBeenCalledTimes(1);
    });
  });

  describe("fire employee confirmation modal", () => {
    let modal: ShallowWrapper;
    beforeEach(() => {
      modal = mkWrapper().find("Confirm");
    });

    it("renders correctly", () => {
      expect(modal).toMatchSnapshot();
    });
    describe("when accepted", () => {
      it("fires the employee", () => {
        const { uid } = state.employee;
        expect(fireEmployee).not.toHaveBeenCalled();

        modal.simulate("yes");

        expect(fireEmployee).toHaveBeenCalledTimes(1);
        expect(fireEmployee).toHaveBeenCalledWith(uid, props.navigation);
      });
    });
    describe("when rejected", () => {
      it("does not fire the employee", () => {
        modal.simulate("no");
        expect(fireEmployee).not.toHaveBeenCalled();
      });
      it("closes the modal", () => {
        expect(closeFireModal).not.toHaveBeenCalled();

        modal.simulate("no");

        expect(closeFireModal).toHaveBeenCalledTimes(1);
      });
    });
  });

  it("sets the page title based on the employeeName", () => {
    const mockGetParam = jest.fn(() => state.employee.employeeName);
    expect(
      navigationOptions({ navigation: { getParam: mockGetParam } }).title,
    ).toMatchSnapshot();
    expect(mockGetParam).toHaveBeenCalledWith("employeeName");
  });

  it("resets the form when unmounted", () => {
    const wrapper = mkWrapper();
    expect(resetForm).not.toHaveBeenCalled();

    wrapper.unmount();

    expect(resetForm).toHaveBeenCalledTimes(1);
  });
});
