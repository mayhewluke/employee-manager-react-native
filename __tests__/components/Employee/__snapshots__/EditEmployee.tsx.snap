// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeList fire employee confirmation modal renders correctly 1`] = `
<Confirm
  isVisible={false}
  onNo={[Function]}
  onYes={[Function]}
  text="Are you sure you want to fire Taylor?"
/>
`;

exports[`EmployeeList sets the page title based on the employeeName 1`] = `"Edit Taylor"`;
